### ISSUE TITLE
Comments/Suggestions/Questions

### ISSUE THREAD

**Issue by tipsy** [bchristenson/example-cassandra-javalin-rest-java#1]
### Thank you !

First of all, I really appreciate you taking the time to do this. Overall it looks solid, but I have a few suggestions and questions.

### Use requestLogger
Instead of: 
```java
.before(context -> {
    context.attribute("start", System.currentTimeMillis());
})
.after(context -> {
    long duration = System.currentTimeMillis()-(Long)context.attribute("start");
    String queryString = (context.queryString() == null)
        ? "" 
        : "?"+context.queryString();
    LOGGER.info(
        context.method()+" "+
        context.path()+queryString+" "+ 
        context.status()+" ("+duration+"ms)");
})
```

You can use:
```java
.requestLogger((context, timeMs) -> {
    String queryString = (context.queryString() == null)
        ? "" 
        : "?"+context.queryString();
    LOGGER.info(
        context.method()+" "+
        context.path()+queryString+" "+ 
        context.status()+" ("+timeMs+"ms)");
})
```
### JSON futures
You have a lot of future-applys:

```java
CompletableFuture<String> future = widgetDao.list(tenantKey, limit, offsetKey)
        .thenApply(widgets -> toJson(ImmutableMap.of("widgets", widgets)));
context.result(future);
```
If the `ImmutableMap.of("widgets", widgets)` bit wasn't required, these could be replaced by the default `json()` method:

```java
context.json(widgetDao.list(tenantKey, limit, offsetKey));
```

This is the method in `Context.kt`:
```kotlin
fun json(future: CompletableFuture<*>): Context {
    val mappingFuture = future.thenApply { JavalinJson.toJson(it) }
    return contentType("application/json").result(mappingFuture)
}
```

It also applies the correct content type, so you wouldn't have to set a default value.

### Route declaration
You have this block:

```java
path("/tenants/:tenantKey", () -> {
    get("/widgets", widgets::list);
    post("/widgets", widgets::create);
    get("/widgets/:key", widgets::retrieve);
    put("/widgets/:key", widgets::update);
    delete("/widgets/:key", widgets::delete);
});
```

It could be written like:

```java
path("/tenants/:tenantKey/widgets", () -> {
    get(widgets::list);
    post(widgets::create);
    path(":key", () -> {
        get(widgets::retrieve);
        put(widgets::update);
        delete(widgets::delete);
    });
});
```

Which would remove a lot of duplicate strings.

### Magic null in startup
```java
public static void main(String[] args) throws Exception {
    start(null);
}
```
Maybe `start(noopStartupCallback)` ? This was a bit confusing before reading the code.

### Formatting of start-method
```
public static void start(
        Runnable afterStartupCallback
) {
```

Why is this not formatted like the rest of the project?
```java
public static void start(afterStartupCallback) {
```

### General
You've wrapped the entire app in a try/catch, wouldn't it be better to isolate the different error cases?
What is the purpose of the `shutdownFuture`?

**Comment by bchristenson** [comment]
> ### Use requestLogger
I didn't know that this existed.  Will definitely switch that over.

> ### JSON futures
The closed-source application that I was mimicking wraps the response in an "envelope" so that other meta information can be included in the response payload.  This probably isn't necessary for everyone, so I will switch to your suggested pattern of:
```java
> context.json(widgetDao.list(tenantKey, limit, offsetKey));
```
> ### Route declaration
This change also makes sense.

> ### Magic null in startup
This makes sense.  I'll switch to:
```java
public static void main(String[] args) throws Exception {
    start(() -> {
        // No-op startup callback; this code would be executed after the server has completed startup
    });
}
```

> ### Formatting of start-method
This just got missed.  Will reformat.

> ### General
> You've wrapped the entire app in a try/catch, wouldn't it be better to isolate the different error cases?

Are you talking specifically about the ExampleApp#start try/catch or the over-simplified Javalin `.exception` handler?  Exception handling is the biggest area in need of cleanup.  The WidgetController should be throwing `NotFoundResponse` exceptions rather than `RuntimeExceptions`, and the `.exception` block should probably be handling more errors (I wasn't sure how much exception "maturity" should be in scope of the sample).

> ### General
> What is the purpose of the `shutdownFuture`?.

The purpose of shutdownFuture is to provide a mechanism to block until a SIGTERM signal is sent to the process (at which point the shutdownFuture is completed and the shutdown process begins).

**Comment by tipsy** [comment]
> Are you talking specifically about the ExampleApp#start try/catch or the over-simplified Javalin .exception handler? Exception handling is the biggest area in need of cleanup. The WidgetController should be throwing NotFoundResponse exceptions rather than RuntimeExceptions, and the .exception block should probably be handling more errors (I wasn't sure how much exception "maturity" should be in scope of the sample).

I was talking about the try-with-resources in `ExampleApp#start`.

Throwing `NotFoundResponse` is a good idea, but I don't really mind general exception mappers in examples. I'm a big fan of keeping things basic, people can expand on the example if they want. 

> The purpose of shutdownFuture is to provide a mechanism to block until a SIGTERM signal is sent to the process (at which point the shutdownFuture is completed and the shutdown process begins).

Typically used for cleanup?

**Comment by bchristenson** [comment]
Yes.  In this case there are open resources (ie the `Cluster` and `Session` object that should be closed explicitly.  In other cases there may be background workers that should "finish" their work without starting new work.

**Comment by tipsy** [comment]
Makes sense!
